### PR TITLE
Implemented power and battery for aovx v series

### PR DESCRIPTION
**Pull request by jcardus**
Implemented power and battery for aovx v series

[AOVX_V Series_Cloud Platform Protocol_V2.9.pdf](https://github.com/user-attachments/files/16038325/AOVX_V.Series_Cloud.Platform.Protocol_V2.9.pdf)

### CODE REVIEW

**Comment by tananaev** [comment]
Encoding should be done in frame encoder, so let's split the PR.

**Comment by jcardus** [comment]
@tananaev please check

File: src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java [hunk 1]
```diff
@@ -772,6 +775,11 @@ private Position decodeLocation(DeviceSession deviceSession, ByteBuf buf) {
                         }
                     }
                     break;
+                case 0xF7:
+                    position.set(Position.KEY_BATTERY, buf.readUnsignedInt() * 0.001);
+                    position.set(Position.KEY_CHARGE, buf.readUnsignedByte() >= 2);
```
Comment (tananaev) [inline src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java:780]: This doesn't look correct. The documentation says that only 2 and 3 indicate charing.

File: src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java [hunk 2]
```diff
@@ -772,6 +775,11 @@ private Position decodeLocation(DeviceSession deviceSession, ByteBuf buf) {
...
+                    position.set(Position.KEY_BATTERY_LEVEL, buf.readUnsignedByte());
```
Comment (tananaev) [inline src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java:781]: It also says these are optional fields. You don't have any checks for that.

File: src/test/java/org/traccar/protocol/HuabaoProtocolDecoderTest.java [hunk 3]
```diff
@@ -227,6 +227,22 @@ public void testDecode() throws Exception {
...
+                Position.KEY_CHARGE, true);
+
+        verifyAttribute(decoder, binary(
+                "7E020000964130564801050048000000001000004F01651E2E02CAEC3802E80000001524062616002101040000000330011F31010EF1040000308CF231414F56585F564C3330302D4C415F48322E315F4547393135554C4141425230334130324D30385F56322E312E305F763130F3011FF400F5080000000000000001F6084008FFFFFFFD0009F70600000E870246F912000F000000010000062A00082406261600184D7E"),
+                Position.KEY_BATTERY_LEVEL, 70);
+
+        verifyAttribute(decoder, binary(
+                "7E020000964130564801050048000000001000004F01651E2E02CAEC3802E80000001524062616002101040000000330011F31010EF1040000308CF231414F56585F564C3330302D4C415F48322E315F4547393135554C4141425230334130324D30385F56322E312E305F763130F3011FF400F5080000000000000001F6084008FFFFFFFD0009F70600000E870246F912000F000000010000062A00082406261600184D7E"),
+                Position.KEY_POWER, 12.428);
+
```
Comment (tananaev) [inline src/test/java/org/traccar/protocol/HuabaoProtocolDecoderTest.java:245]: I don't think we need so many test cases. Two is probably enough. One to test `0xF1` and one to test last parameter from `0xF7`.

File: src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java
[the remark this answers is not in the document]
Comment (tananaev) [reply src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java:1]: You have check by model.

[the remark this answers is not in the document]
Comment (jcardus) [reply src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java:1]: device model comes in device registration message, do you know if it's the same with huabao?

[the remark this answers is not in the document]
Comment (jcardus) [reply src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java:1]: looking at [this](https://www.traccar.org/protocol/5015-huabao/Communication%20Protocol%20of%20VL502%20%20V1.0.7_eng.pdf), looks like not.

[the remark this answers is not in the document]
Comment (tananaev) [reply src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java:1]: Device model is something you can configure on the server side.

[the remark this answers is not in the document]
Comment (jcardus) [reply src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java:1]: Yes I know, I prefer to detect it automatically

File: src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java [hunk 4]
```diff
@@ -772,6 +775,24 @@ private Position decodeLocation(DeviceSession deviceSession, ByteBuf buf) {
                         }
                     }
                     break;
+                case 0xF7:
+                    position.set(Position.KEY_BATTERY, buf.readUnsignedInt() * 0.001);
+                    if (length > 4) {
```
Comment (tananaev) [inline src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java:780]:
```suggestion
                    if (length >= 5) {
```

File: src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java [hunk 5]
```diff
@@ -772,6 +775,24 @@ private Position decodeLocation(DeviceSession deviceSession, ByteBuf buf) {
...
+                            case 1:
+                                position.set(Position.KEY_CHARGE, false);
+                                break;
+                            case 2:
+                            case 3:
+                                position.set(Position.KEY_CHARGE, true);
+                            default:
+                        }
+                    }
+                    if (length > 5) {
```
Comment (tananaev) [inline src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java:792]:
```suggestion
                    if (length >= 6) {
```

File: src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java [hunk 6]
```diff
@@ -772,6 +775,24 @@ private Position decodeLocation(DeviceSession deviceSession, ByteBuf buf) {
                         }
                     }
                     break;
+                case 0xF7:
+                    position.set(Position.KEY_BATTERY, buf.readUnsignedInt() * 0.001);
+                    if (length > 4) {
+                        short batteryStatus = buf.readUnsignedByte();
+                        switch (batteryStatus) {
+                            case 1:
+                                position.set(Position.KEY_CHARGE, false);
```
Comment (tananaev) [inline src/main/java/org/traccar/protocol/HuabaoProtocolDecoder.java:784]: We don't store `false` values for charge.

**Pull request merged**